src/hotspot/share/opto/divnode.cpp

```diff
@@ -1203,72 +1203,77 @@ Node *ModINode::Ideal(PhaseGVN *phase, bool can_reshape) {
+static const Type* mod_value(const PhaseGVN* phase, const Node* in1, const Node* in2, const BasicType bt, const Type* bottom) {
   // Either input is TOP ==> the result is TOP
-  const Type *t1 = phase->type( in(1) );
-  const Type *t2 = phase->type( in(2) );
-  if( t1 == Type::TOP ) return Type::TOP;
-  if( t2 == Type::TOP ) return Type::TOP;
+  const Type* t1 = phase->type(in1);
+  const Type* t2 = phase->type(in2);
+  if (t1 == Type::TOP) return Type::TOP;
+  if (t2 == Type::TOP) return Type::TOP;
 
   // We always generate the dynamic check for 0.
   // 0 MOD X is 0
-  if( t1 == TypeInt::ZERO ) return TypeInt::ZERO;
+  if (t1 == TypeInteger::zero(bt)) return t1;
+
   // X MOD X is 0
-  if (in(1) == in(2)) {
-    return TypeInt::ZERO;
+  if (in1 == in2) {
+    return TypeInteger::zero(bt);
   }
 
-  const TypeInt *i1 = t1->isa_int();
-  const TypeInt *i2 = t2->isa_int();
-
-  if (i1 == nullptr || i2 == nullptr) {
-    return bottom_type();
-  }
   // Mod by zero?  Throw exception at runtime!
-  if (i2->is_con() && i2->get_con() == 0) {
+  if (t2 == TypeInteger::zero(bt)) {
     return TypeInt::TOP;
   }
+
+  const TypeInteger* i1 = t1->isa_integer(bt);
+  const TypeInteger* i2 = t2->isa_integer(bt);
+  if (i1 == nullptr || i2 == nullptr) {
+    return bottom;
+  }
   if (i1->is_con() && i2->is_con()) {
     // We must be modulo'ing 2 int constants.
-    // Check for min_jint % '-1', result is defined to be '0'.
-    if (i1->get_con() == min_jint && i2->get_con() == -1) {
-      return TypeInt::ZERO;
+    // Check for min_jlong % '-1', result is defined to be '0'
+    // We don't need to check for min_jint % '-1' as its result is defined when using jlong.
+    if (i1->get_con_as_long(bt) == min_jlong && i2->get_con_as_long(bt) == -1) {
+      return TypeInteger::zero(bt);
     }
-    return TypeInt::make(i1->get_con() % i2->get_con());
+    return TypeInteger::make(i1->get_con_as_long(bt) % i2->get_con_as_long(bt), bt);
   }
-
-  // The magnitude of the divisor is in range [1, 2^31].
+  // The magnitude of the divisor is in range [1, 2^63].
   // We know it isn't 0 as we handled that above.
   // That means at least one value is nonzero, so its absolute value is bigger than zero.
-  juint divisor_magnitude = MAX2(g_uabs(i2->_lo), g_uabs(i2->_hi));
-  // JVMS irem bytecode: "the magnitude of the result is always less than the magnitude of the divisor"
-  // "less than" means we can subtract 1 to get an inclusive upper bound in [0, 2^31-1]
-  jint hi = static_cast<jint>(divisor_magnitude - 1);
-  jint lo = -hi;
-  // JVMS irem bytecode: "the result of the remainder operation can be negative only if the dividend
+  julong divisor_magnitude = MAX2(g_uabs(i2->lo_as_long()), g_uabs(i2->hi_as_long()));
+  // JVMS lrem bytecode: "the magnitude of the result is always less than the magnitude of the divisor"
+  // "less than" means we can subtract 1 to get an inclusive upper bound in [0, 2^63-1]
+  jlong hi = static_cast<jlong>(divisor_magnitude - 1);
```

Comment:
Hmm, this also looks confusing for the `T_INT` case. What about `-5`, does that then not become `max_julong - 5`, but it should have been `max_juint - 1`?

Reply:
We use `g_uabs()` to get the absolute value, that should't exceed 2^31 for int values (i.e., `g_uabs(min_jint) == 2^31`). So we should get into the right range here again. But I guess I can expand the comment to better explain that part.

Reply:
Yes that's correct, and it should still work due to how negation works for negative inputs.

Reply:
I think I was quite a bit confused here. Sounds good to me now though :)